<a href="https://colab.research.google.com/github/heisenberg304/Housing-Prices-prediction/blob/develop/Modeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Modeling

│

├── Dummy Regressor


├── Linear Regression


├── Ridge Regression


├── Lasso Regression


├── Decision Tree Regressor


├── Random Forest Regressor


├── Gradient Boosting


└── XGBoost

#LOAD DATA

In [20]:
%%writefile data_loading.py
import pandas as pd
from sklearn.model_selection import train_test_split

#cargar datos
# data train no filled
train_not_filled = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/HP_train_not_filled_encoding.csv")

# data train filled
train_filled = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/HP_train_filled_encoding.csv")

#x,y
y = train_filled["SalePrice"]
x = train_filled.drop("SalePrice", axis=1)

#x,y
y_nf = train_not_filled["SalePrice"]
x_nf = train_not_filled.drop("SalePrice", axis=1)


def split_dataset(x, y):  #split 80/20
  x_train, x_test, y_train, y_test = train_test_split(
      x,
      y,
      test_size = 0.2,
      random_state=42
  )
  return x_train, x_test, y_train, y_test
x_train, x_test, y_train, y_test = split_dataset(x, y)
x_train_nf, x_test_nf, y_train_nf, y_test_nf = split_dataset(x_nf, y_nf)





Writing data_loading.py


In [21]:
%%writefile modeling_def.py
#from data_loading import x_train, y_train, x_test, y_test
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score

#entrenar modelo
def entrenar_model(model, x, y):
  return model.fit(x,y)

#predicciones del modelo
def predict_model(model, x):
  return model.predict(x)

#metricas del modelo
def evaluate_model(y_true, y_pred):
  MSE = mean_squared_error(y_true, y_pred)
  RMSE = root_mean_squared_error(y_true, y_pred)
  MAE = mean_absolute_error(y_true, y_pred)
  R2 = r2_score(y_true, y_pred)
  return MSE, RMSE, MAE, R2

#

Writing modeling_def.py


#Baseline
   - Dummy

In [44]:
#Dummy Regressor
from data_loading import x_train, y_train, x_test, y_test
from sklearn.dummy import DummyRegressor
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = DummyRegressor()

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#

 Root Mean Squared Error: 16603.77832552691
 Mean Absolute Error: 12084.41645179763
 R2 Score: -0.01106892316122332 
Mean Squared Error: 275685454.6832372


#Linear Models
   - Linear
   - Ridge
   - Lasso

In [26]:
#Linear Regression
from sklearn.linear_model import LinearRegression
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = LinearRegression()

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#

 Root Mean Squared Error: 574883.4453389066
 Mean Absolute Error: 566946.2718963341
 R2 Score: -1211.0666842015441 
Mean Squared Error: 330490975724.73157


In [34]:
#Ridge Regression
from sklearn.linear_model import Ridge
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = Ridge(alpha=10)

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#0.1 = Root Mean Squared Error: 208001.0122151831
#Mean Absolute Error: 195930.62899921017
#R2 Score: -5.471553798013038

#1 o default = Root Mean Squared Error: 87241.37609762137
#Mean Absolute Error: 74944.73677973443
#R2 Score: -0.2595207187564703

#10 = Root Mean Squared Error: 74146.36376031929
#Mean Absolute Error: 58564.711397575076
#R2 Score: 0.06858399347197763

#100 = Root Mean Squared Error: 71879.01208662726
#Mean Absolute Error: 57324.15833066368
#R2 Score: 0.07754967887425646


 Root Mean Squared Error: 74146.36376031929
 Mean Absolute Error: 58564.711397575076
 R2 Score: -19.162604149071317 
Mean Squared Error: 5497683258.87759


In [38]:
#Lasso Regression
from sklearn.linear_model import Lasso
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = Lasso(alpha=1000)

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#0.1 = Root Mean Squared Error: 567330.3876842158
#Mean Absolute Error: 559422.7521880025
#R2 Score: -30.39038579444292

#1 o default = Root Mean Squared Error: 536897.3344515655
#Mean Absolute Error: 528918.0324155539
#R2 Score: -28.422579301207755

#10 = Root Mean Squared Error: 251061.15564594237
#Mean Absolute Error: 239611.16133609207
#R2 Score: -7.629800405653116

#100 = Root Mean Squared Error: 73633.8292292412
#Mean Absolute Error: 57559.13697392844
#R2 Score: 0.07560295838079045

 Root Mean Squared Error: 69238.72666516973
 Mean Absolute Error: 55187.497508908804
 R2 Score: -16.581869553031638 
Mean Squared Error: 4794001270.214085


#Tree Models
   - Decision Tree Regressor
   - Random Forest Regressor

In [39]:
#Decision Tree Regressor
from sklearn.tree import DecisionTreeRegressor
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = DecisionTreeRegressor(max_depth=4,          #3-5
                              min_samples_split=3,  #2-4
                              min_samples_leaf=1,   #1-2
                              random_state=7
)

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#defaul =  Root Mean Squared Error: 78681.31931644102
#Mean Absolute Error: 60402.01052983383
#R2 Score: 0.0559027093532396


#buscando =   Root Mean Squared Error: 67716.40511697781
# Mean Absolute Error: 50276.14044378846
# R2 Score: 0.06566950898019164



 Root Mean Squared Error: 67716.40511697781
 Mean Absolute Error: 50276.14044378846
 R2 Score: -15.817239059584363 
Mean Squared Error: 4585511521.966658


In [40]:
#Random Forest Regressor
from sklearn.ensemble import RandomForestRegressor
from data_loading import x_train, y_train, x_test, y_test
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = RandomForestRegressor(n_estimators= 100, #50-100
                              max_depth= 4,      #4-5
                              min_samples_leaf=4,#4-5
                              min_samples_split=10, #3-10
                              max_features="sqrt",
                              random_state= 3
)
#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#default =  Root Mean Squared Error: 69594.30387177203
#Mean Absolute Error: 53363.752645650944
#R2 Score: 0.07832547636565712

#4 =   Root Mean Squared Error: 51659.79870246149
# Mean Absolute Error: 41535.08111895361
# R2 Score: 0.056789290070577536

 Root Mean Squared Error: 51659.79870246149
 Mean Absolute Error: 41535.08111895361
 R2 Score: -8.787512458863482 
Mean Squared Error: 2668734801.978842


#Boosting Models
   - Gradient Boosting Regressor
   - XGBoost Regressor

In [42]:
#Gradient Boosting Regressor
from sklearn.ensemble import GradientBoostingRegressor
from data_loading import x_train, y_train, x_test, y_test, y_train_nf, x_train_nf, y_test_nf, x_test_nf
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = GradientBoostingRegressor(#learning_rate=0.01,
                                  n_estimators=150,
                                  #max_depth=,
                                  #subsample=,
                                  random_state=5
)
#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \n Mean Squared Error: {MSE}')

#default = Root Mean Squared Error: 73135.35367116224
# Mean Absolute Error: 55000.79698477913
# R2 Score: 0.08330796373524485

#

 Root Mean Squared Error: 73470.9705191755
 Mean Absolute Error: 55092.33320353205
 R2 Score: -18.796958022277586 
 Mean Squared Error: 5397983509.029557


In [43]:
from xgboost import XGBRegressor
import data_loading
import importlib
importlib.reload(data_loading)
from data_loading import x_train, y_train, x_test, y_test, y_train_nf, x_train_nf, y_test_nf, x_test_nf
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = XGBRegressor()

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test, predict)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

###################################################################################################################################################

#entrenar modelo
modelo_entrenado_nf = entrenar_model(model, x_train_nf, y_train_nf)

#predicciones del modelo
predict_nf = predict_model(modelo_entrenado_nf, x_test_nf)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(y_test_nf, predict_nf)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

 Root Mean Squared Error: 73397.0625
 Mean Absolute Error: 55032.90625
 R2 Score: -18.75714683532715 
Mean Squared Error: 5387128320.0
 Root Mean Squared Error: 73776.9296875
 Mean Absolute Error: 54509.6875
 R2 Score: -18.96218490600586 
Mean Squared Error: 5443035136.0


#Tabla comparativa
---

| Modelo | RMSE ↓ | MAE ↓ | R² ↑ | MSE |
|---------|--------:|-------:|------:|:------:|
| Dummy |16603.778 |12084.416 |-3.255 |275685454.683 |
| Linear |574883.445 |566946.272 |-30.7437 |330490975724.73 |
| Ridge | | | | |
| Lasso | | | | |
| Decision Tree | | | | |
| Random Forest | | | | |
| Gradient Boosting | | | | |
| XGBoost | | | | |
| XGBoost_Not_Filled | | | | |
---

